torch/functional.py

```diff
@@ -294,10 +294,15 @@ def einsum(equation, *operands):
 """
     if not torch.jit.is_scripting():
         if any(type(t) is not Tensor for t in operands) and has_torch_function(operands):
-            return handle_torch_function(einsum, operands, *operands)
+            return handle_torch_function(einsum, operands, equation, *operands)
+
     if len(operands) == 1 and isinstance(operands[0], (list, tuple)):
         # the old interface of passing the operands as one list argument
         operands = operands[0]
+        # recurse incase operands contains value that has torch function
+        # in the original implementation this line is omitted
+        return einsum(equation, *operands)
```

Comment:
This shouldn’t be necessary IIUC. `operands[0]` will already Dispatch above.

Reply:
No that's not the case. See the `if any(type(t) ....` line that guards `handle_torch_function`. Dispatching happens on a sequence of `Tensor`s or anything with a `__torch_function__` method, not with a sequence of "objects with `__torch_function__` method".

Without this line one gets:
```
  File "/home/rgommers/code/pytorch/torch/functional.py", line 306, in einsum
    return _VF.einsum(equation, operands)
TypeError: expected Tensor as element 0 in argument 1, but got Wrapper
```

Reply:
Ah, understood.